### PR TITLE
Travis: add windows CI

### PR DESCRIPTION
**Pull request by matthiasdiener**
Run Travis CI tests on a Windows VM. 

Due to Travis time limitations (90 min) and the extremely long build/test time, runs only Charm++ tests.

### CODE REVIEW

**Comment by evan-charmworks** [comment]
Would it work to test a CMake build on Windows using Travis instead?

**Comment by matthiasdiener** [comment]
> For Travis, would it work to test a CMake build on Windows instead?

No, we still need autoconf for hwloc.

**Comment by evan-charmworks** [comment]
> > For Travis, would it work to test a CMake build on Windows instead?
> 
> No, we still need autoconf for hwloc.

#2661 should fix that.

**Comment by evan-charmworks** [comment]
Maybe `export MAKE=make` will get MSYS2 to use make instead of mingw32-make. If that doesn't work or has some drawback, try installing `mingw-w64-x86_64-make` with pacman.

File: .travis.yml [hunk 1]
```diff
@@ -1,44 +1,59 @@
...
- #   - ./build LIBS netlrts-win-x86_64 -j2 -k --without-romio --with-production --enable-error-checking --enable-lbuserdata -g || exit 1
- #   - make -C netlrts-win-x86_64/tmp -k -j2 all-test-tests all-test-examples OPTS="-g"
- #   - make -C netlrts-win-x86_64/tmp -k test-tests test-examples TESTOPTS="++local"
+
+  - os: windows
+    name: "NetLRTS-Win"
+    script:
+    - $msys2 ./build AMPI netlrts-win-x86_64 gcc -j3 --without-romio --with-production --enable-error-checking --enable-lbuserdata -g || exit 1
+    - $msys2 make -C netlrts-darwin-x86_64/tmp -j3 all-test-tests all-test-examples OPTS="-g"
+    - $msys2 make -C netlrts-darwin-x86_64/tmp test-tests test-examples TESTOPTS="++local"
```
Comment (evan-charmworks) [inline .travis.yml]: `netlrts-darwin-x86_64` should be `netlrts-win-x86_64-gcc`

Comment (matthiasdiener) [reply .travis.yml]: fixed, thanks.

File: .travis.yml [hunk 2]
```diff
@@ -1,44 +1,59 @@
...
- #   - ./build LIBS netlrts-win-x86_64 -j2 -k --without-romio --with-production --enable-error-checking --enable-lbuserdata -g || exit 1
- #   - make -C netlrts-win-x86_64/tmp -k -j2 all-test-tests all-test-examples OPTS="-g"
- #   - make -C netlrts-win-x86_64/tmp -k test-tests test-examples TESTOPTS="++local"
+
+  - os: windows
+    name: "NetLRTS-Win"
+    script:
+    - $msys2 ./build LIBS netlrts-win-x86_64 gcc -j3 --without-romio --with-production --enable-error-checking --enable-lbuserdata -g || exit 1
+    - $msys2 make -C netlrts-win-x86_64-gcc/tmp -j3 all-test-tests all-test-examples OPTS="-g"
+    - $msys2 make -C netlrts-win-x86_64-gcc/tmp test-tests test-examples TESTOPTS="++local"
```
Comment (evan-charmworks) [inline .travis.yml]: Should these `$msys2`s be `$mingw64`s?

File: .travis.yml [hunk 3]
```diff
@@ -1,44 +1,59 @@
...
- #   name: "NetLRTS-Win"
- #   script:
- #   - choco install -y msys2
- #   - pacman -S autoconf automake
- #   - ./build LIBS netlrts-win-x86_64 -j2 -k --without-romio --with-production --enable-error-checking --enable-lbuserdata -g || exit 1
- #   - make -C netlrts-win-x86_64/tmp -k -j2 all-test-tests all-test-examples OPTS="-g"
- #   - make -C netlrts-win-x86_64/tmp -k test-tests test-examples TESTOPTS="++local"
+
+  - os: windows
+    name: "NetLRTS-Win"
```
Comment (evan-charmworks) [inline .travis.yml]: Would recommend naming this something like "NetLRTS-Win GCC" to distinguish it from the default MSVC builds

File: .travis.yml (hunk 2 above)
Comment (matthiasdiener) [reply .travis.yml]: If I understood https://github.com/travis-ci/docs-travis-ci-com/blob/686fbde7444f054808ecb00d9719ac207876527a/user/reference/windows.md correctly, then it should be $msys2. The current version install autoconf/automake correctly, and they appear to work fine. The build later fails due to a seemingly unrelated issue.

File: .travis.yml (hunk 3 above)
Comment (matthiasdiener) [reply .travis.yml]: fixed.

**Comment by matthiasdiener** [comment]
I was able to install autoconf in Travis. The current build fails on Windows/gcc with the following errors:

```
topology.c:150:17: error: ‘_SC_NPROCESSORS_CONF’ undeclared (first use in this function)
  150 |     n = sysconf(_SC_NPROCESSORS_CONF);
      |                 ^~~~~~~~~~~~~~~~~~~~
topology.c:150:17: note: each undeclared identifier is reported only once for each function it appears in
topology.c:162:15: error: ‘_SC_NPROCESSORS_ONLN’ undeclared (first use in this function)
  162 |   n = sysconf(_SC_NPROCESSORS_ONLN);
      |               ^~~~~~~~~~~~~~~~~~~~
make[2]: *** [Makefile:877: topology.lo] Error 1
```

File: .travis.yml (hunk 2 above)
Comment (evan-charmworks) [reply .travis.yml]: I believe switching these to `$mingw64` will fix the current build error.

Comment (matthiasdiener) [reply .travis.yml]: ok, will try.

Comment (evan-charmworks) [reply .travis.yml]: Confirmed, the build made it past topology.c.

**Comment by evan-charmworks** [comment]
The build got much further but failed among the LIBS:

```
make[2]: Leaving directory '/c/Users/travis/build/UIUC-PPL/charm/netlrts-win-x86_64-gcc/tmp/libs/ck-libs/metis/GKlib'
make[1]: Leaving directory '/c/Users/travis/build/UIUC-PPL/charm/netlrts-win-x86_64-gcc/tmp/libs/ck-libs/metis'
In file included from GKlib.h:35,
                 from b64.c:20:
gk_arch.h:46:12: fatal error: sys/resource.h: No such file or directory
   46 |   #include <sys/resource.h>
      |            ^~~~~~~~~~~~~~~~
compilation terminated.
```

This error is consistent with my experience building Charm on MSYS2. If we want to build the tests, we'll need to fix building LIBS. Fortunately, I don't think there are any further issues to work out with Travis integration itself.

**Comment by matthiasdiener** [comment]
> The build got much further but failed among the LIBS:
> 
> ```
> make[2]: Leaving directory '/c/Users/travis/build/UIUC-PPL/charm/netlrts-win-x86_64-gcc/tmp/libs/ck-libs/metis/GKlib'
> make[1]: Leaving directory '/c/Users/travis/build/UIUC-PPL/charm/netlrts-win-x86_64-gcc/tmp/libs/ck-libs/metis'
> In file included from GKlib.h:35,
>                  from b64.c:20:
> gk_arch.h:46:12: fatal error: sys/resource.h: No such file or directory
>    46 |   #include <sys/resource.h>
>       |            ^~~~~~~~~~~~~~~~
> compilation terminated.
> ```
> 
> This error is consistent with my experience building Charm on MSYS2. If we want to build the tests, we'll need to fix building LIBS. Fortunately, I don't think there are any further issues to work out with Travis integration itself.

can we fix this by building all-test ?

**Comment by evan-charmworks** [comment]
> can we fix this by building all-test ?

all-test depends on LIBS.

**Comment by matthiasdiener** [comment]
> The build got much further but failed among the LIBS:
> 
> ```
> make[2]: Leaving directory '/c/Users/travis/build/UIUC-PPL/charm/netlrts-win-x86_64-gcc/tmp/libs/ck-libs/metis/GKlib'
> make[1]: Leaving directory '/c/Users/travis/build/UIUC-PPL/charm/netlrts-win-x86_64-gcc/tmp/libs/ck-libs/metis'
> In file included from GKlib.h:35,
>                  from b64.c:20:
> gk_arch.h:46:12: fatal error: sys/resource.h: No such file or directory
>    46 |   #include <sys/resource.h>
>       |            ^~~~~~~~~~~~~~~~
> compilation terminated.
> ```
> 
> This error is consistent with my experience building Charm on MSYS2. If we want to build the tests, we'll need to fix building LIBS. Fortunately, I don't think there are any further issues to work out with Travis integration itself.

Mingw-metis for msys has this patch: https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-metis/0001-mingw-w64-does-not-have-sys-resource-h.patch 
Should we apply something like that as well?

**Comment by matthiasdiener** [comment]
We should probably switch to msvc. Even on dexterity, gcc doesn't compile.

File: .travis.yml [hunk 4]
```diff
@@ -1,44 +1,67 @@
...
+    case $TRAVIS_OS_NAME in
+      windows)
+        [[ ! -f C:/tools/msys64/msys2_shell.cmd ]] && rm -rf C:/tools/msys64
+        choco upgrade --no-progress -y msys2
+        export msys2='cmd //C RefreshEnv.cmd '
+        export msys2+='& set MSYS=winsymlinks:nativestrict '
+        export msys2+='& C:\\tools\\msys64\\msys2_shell.cmd -defterm -no-start'
+        export mingw64="$msys2 -mingw64 -full-path -here -c \$\* --"
+        export msys2+=" -msys2 -c \$\* --"
+        $msys2 pacman --sync --noconfirm --needed mingw-w64-x86_64-toolchain autoconf libtool automake make autoconf-archive pkg-config
```
Comment (evan-charmworks) [inline .travis.yml]: If we're using MSVC now, I _think_ we don't need `mingw-w64-x86_64-toolchain` here.

File: .travis.yml [hunk 5]
```diff
@@ -1,44 +1,67 @@
...
- #   name: "NetLRTS-Win"
- #   script:
- #   - choco install -y msys2
- #   - pacman -S autoconf automake
- #   - ./build LIBS netlrts-win-x86_64 -j2 -k --without-romio --with-production --enable-error-checking --enable-lbuserdata -g || exit 1
- #   - make -C netlrts-win-x86_64/tmp -k -j2 all-test-tests all-test-examples OPTS="-g"
- #   - make -C netlrts-win-x86_64/tmp -k test-tests test-examples TESTOPTS="++local"
+
+  - os: windows
+    name: "NetLRTS-Win GCC"
```
Comment (evan-charmworks) [inline .travis.yml]: No longer GCC

File: .travis.yml [hunk 6]
```diff
@@ -1,44 +1,67 @@
...
+    - export PATH="/c/Program Files (x86)/Microsoft Visual Studio 14.0/VC/BIN/amd64:$PATH"
+    - cl.exe
+    - $mingw64 echo $PATH
+    - $mingw64 cl.exe
+    - $mingw64 ./build all-test netlrts-win-x86_64 -j3 --without-romio --with-production --enable-error-checking --enable-lbuserdata -g || exit 1
+    - $mingw64 make -C netlrts-win-x86_64-gcc/tmp -j3 all-test-tests all-test-examples OPTS="-g"
+    - $mingw64 make -C netlrts-win-x86_64-gcc/tmp test-tests test-examples TESTOPTS="++local"
```
Comment (evan-charmworks) [inline .travis.yml]: Likewise, no longer GCC

File: .travis.yml [hunk 7]
```diff
@@ -1,44 +1,67 @@
...
+    - export LIBPATH='C:\Windows\Microsoft.NET\Framework64\v4.0.30319;C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\LIB\amd64;C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\ATLMFC\LIB\amd64;C:\Program Files (x86)\Windows Kits\8.1\References\CommonConfiguration\Neutral;\Microsoft.VCLibs\14.0\References\CommonConfiguration\neutral;C:\Windows\Microsoft.NET\Framework64\v4.0.30319;C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\LIB\amd64;C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\ATLMFC\LIB\amd64;C:\Program Files (x86)\Windows Kits\8.1\References\CommonConfiguration\Neutral;\Microsoft.VCLibs\14.0\References\CommonConfiguration\neutral;C:\Program Files (x86)\Microsoft SDKs\MPI\Lib\x64'
+    - export PATH="/c/Program Files (x86)/Microsoft Visual Studio 14.0/VC/BIN/amd64:$PATH"
```
Comment (evan-charmworks) [inline .travis.yml]: Maybe change this to `msvc="$mingw64 export PATH=\"/c/Program Files (x86)/Microsoft Visual Studio 14.0/VC/BIN/amd64:\$PATH\" &&"` and change the below `$mingw64`s to `$msvc`s

Comment (evan-charmworks) [reply .travis.yml]: Might need additional escaping though.

**Comment by evan-charmworks** [comment]
The build begins now but is unable to complete due to disk space.

**Comment by matthiasdiener** [comment]
> The build begins now but is unable to complete due to disk space.

Removing the `-g` flag should reduce disk usage considerably.

**Comment by matthiasdiener** [comment]
This currently hangs during `msys2` installation:

```
hocolatey v0.10.15
Upgrading the following packages:
msys2
By upgrading you accept licenses for the packages.
msys2 is not installed. Installing...
Progress: Downloading msys2 20190524.0.0.20191030... 100%
Progress: Downloading msys2 20190524.0.0.20191030... 100%
msys2 v20190524.0.0.20191030 [Approved]
msys2 package files upgrade completed. Performing other installation steps.
'C:\tools\msys64' already exists and will only be updated.
Invoking first run to setup things like bash profile, gpg etc...
Invoking msys2 shell command: -defterm -no-start -c "ps -ef | grep '[?]' | awk '{print $2}' | xargs -r kill"
```

File: tests/charm++/jacobi3d-sdag/jacobi3d.C [hunk 8]
```diff
@@ -171,7 +171,7 @@ class Main : public CBase_Main {
 
       TopoManager tmgr;
       CkArray *jarr = array.ckLocalBranch();
-      int jmap[num_chare_x][num_chare_y][num_chare_z];
+      std::vector<std::vector<std::vector<int>>> jmap(num_chare_x, std::vector<std::vector<int> >(num_chare_y,std::vector <int>(num_chare_z)));
```
Comment (matthiasdiener) [inline tests/charm++/jacobi3d-sdag/jacobi3d.C]: This change and the change below are replicated from #2670 to make testing easier.

**Comment by matthiasdiener** [comment]
The current version runs fine on Travis (although slowly). Ready for review.

**Comment by matthiasdiener** [comment]
> One concern I have with merging this is that since Windows verification takes 80 minutes, it slows down how fast we can merge PRs that are approved. That's not a dealbreaker though. Maybe we can make use of the change to bash to speed up the build process.

I think megatest is what takes the most time on the Windows VM. Reducing the number of iterations in immediatering by 10x reduces the CI time to a more reasonable 50 min.

**Comment by matthiasdiener** [comment]
> > One concern I have with merging this is that since Windows verification takes 80 minutes, it slows down how fast we can merge PRs that are approved. That's not a dealbreaker though. Maybe we can make use of the change to bash to speed up the build process.
> 
> I think megatest is what takes the most time on the Windows VM. Reducing the number of iterations in immediatering by 10x reduces the CI time to a more reasonable 50 min.

Just to clarify, I added this reduction in the number of iterations to this patch. Can this stay here or should it be a separate patch?

**Comment by matthiasdiener** [comment]
Needs another review.

**Pull request merged**